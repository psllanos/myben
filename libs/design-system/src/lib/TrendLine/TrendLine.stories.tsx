import type { Story, Meta } from '@storybook/react'
import type { TrendLineProps } from '.'

import TrendLine from './TrendLine'

const data = Array.from({ length: 30 }, () => Math.random() * 10000).map((value, index) => ({
    key: index,
    value,
}))

export default {
    title: 'Components/TrendLine',
    component: TrendLine,
    parameters: { controls: { exclude: ['className'] } },
    args: {
        inverted: false,
        data,
    },
} as Meta

export const Base: Story<TrendLineProps> = (props) => (
    <div className="w-16 h-4">
        <TrendLine {...props} />
    </div>
)
