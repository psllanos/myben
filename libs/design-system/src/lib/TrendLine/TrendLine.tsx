import { useMemo } from 'react'
import { ParentSize } from '@visx/responsive'
import { scaleBand, scaleLinear } from '@visx/scale'
import { Group } from '@visx/group'
import { Area, Circle } from '@visx/shape'
import classNames from 'classnames'

const CIRCLE_RADIUS = 1

export interface TrendLineProps extends React.HTMLAttributes<HTMLDivElement> {
    data: {
        key: { toString(): string }
        value: number
    }[]
    /** Whether a negative trendline should be treated positively (teal vs. red) */
    inverted?: boolean
}

export default function TrendLine({
    data,
    className,
    inverted = false,
    ...rest
}: TrendLineProps): JSX.Element | null {
    const xScale = useMemo(
        () =>
            scaleBand<string>({
                domain: data.map(({ key }) => key.toString()),
                round: false,
            }),
        [data]
    )
    const yScale = useMemo(() => {
        const yDomain = data.map(({ value }) => value)

        return scaleLinear<number>({
            domain: [Math.min(...yDomain), Math.max(...yDomain)],
            round: true,
        })
    }, [data])

    const change = data[data.length - 1].value - data[0].value

    let isPositive = change > 0
    if (inverted) isPositive = !isPositive

    return data.length > 0 ? (
        <ParentSize className={classNames('relative', className)} {...rest}>
            {({ width, height }) => {
                xScale.range([0, width - CIRCLE_RADIUS * 2])
                yScale.range([height - CIRCLE_RADIUS * 2, 0])

                return (
                    <svg {...{ width, height }}>
                        <Group left={CIRCLE_RADIUS} top={CIRCLE_RADIUS}>
                            <>
                                <Area
                                    data={data}
                                    x={({ key }) => xScale(key.toString()) || 0}
                                    y={({ value }) => yScale(value)}
                                    className={
                                        change === 0
                                            ? 'text-gray-200'
                                            : isPositive
                                            ? 'text-teal'
                                            : 'text-red'
                                    }
                                    stroke="currentColor"
                                    strokeWidth={1}
                                    strokeLinejoin="round"
                                />
                                <Circle
                                    r={CIRCLE_RADIUS}
                                    cx={xScale(data[data.length - 1].key.toString())}
                                    cy={yScale(data[data.length - 1].value)}
                                    className={
                                        change === 0
                                            ? 'text-gray-200'
                                            : isPositive
                                            ? 'text-teal'
                                            : 'text-red'
                                    }
                                    fill="currentColor"
                                />
                            </>
                        </Group>
                    </svg>
                )
            }}
        </ParentSize>
    ) : null
}
