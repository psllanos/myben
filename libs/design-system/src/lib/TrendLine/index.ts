export { default as TrendLine } from './TrendLine'
export type { TrendLineProps } from './TrendLine'
