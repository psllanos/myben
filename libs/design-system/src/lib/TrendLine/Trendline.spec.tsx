import { render } from '@testing-library/react'
import TrendLine from './TrendLine'

describe('TrendLine', () => {
    describe('when rendered with a positive trendline', () => {
        it('should render correcly', () => {
            const component = render(
                <TrendLine
                    data={[
                        { key: 1, value: 1 },
                        { key: 2, value: 2 },
                        { key: 3, value: 3 },
                    ]}
                />
            )

            expect(component).toMatchSnapshot()
        })
    })

    describe('when rendered with a negative trendline', () => {
        it('should render correcly', () => {
            const component = render(
                <TrendLine
                    data={[
                        { key: 3, value: 3 },
                        { key: 2, value: 2 },
                        { key: 1, value: 1 },
                    ]}
                />
            )

            expect(component).toMatchSnapshot()
        })
    })

    describe('when rendered with a neutral trendline', () => {
        it('should render correcly', () => {
            const component = render(
                <TrendLine
                    data={[
                        { key: 1, value: 1 },
                        { key: 2, value: 2 },
                        { key: 3, value: 1 },
                    ]}
                />
            )

            expect(component).toMatchSnapshot()
        })
    })
})
