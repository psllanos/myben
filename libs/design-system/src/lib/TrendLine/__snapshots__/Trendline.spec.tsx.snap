// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrendLine when rendered with a negative trendline should render correcly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
        style="width: 100%; height: 100%;"
      >
        <svg
          height="0"
          width="0"
        >
          <g
            class="visx-group"
            transform="translate(1, 1)"
          >
            <path
              class="visx-area text-red"
              d="M-0.6666666666666667,0L-1.3333333333333335,-1L-2,-2L-2,-2L-1.3333333333333335,-1L-0.6666666666666667,0Z"
              stroke="currentColor"
              stroke-linejoin="round"
              stroke-width="1"
            />
            <circle
              class="visx-circle text-red"
              cx="-2"
              cy="-2"
              fill="currentColor"
              r="1"
            />
          </g>
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
      style="width: 100%; height: 100%;"
    >
      <svg
        height="0"
        width="0"
      >
        <g
          class="visx-group"
          transform="translate(1, 1)"
        >
          <path
            class="visx-area text-red"
            d="M-0.6666666666666667,0L-1.3333333333333335,-1L-2,-2L-2,-2L-1.3333333333333335,-1L-0.6666666666666667,0Z"
            stroke="currentColor"
            stroke-linejoin="round"
            stroke-width="1"
          />
          <circle
            class="visx-circle text-red"
            cx="-2"
            cy="-2"
            fill="currentColor"
            r="1"
          />
        </g>
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TrendLine when rendered with a neutral trendline should render correcly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
        style="width: 100%; height: 100%;"
      >
        <svg
          height="0"
          width="0"
        >
          <g
            class="visx-group"
            transform="translate(1, 1)"
          >
            <path
              class="visx-area text-gray-200"
              d="M-0.6666666666666667,-2L-1.3333333333333335,0L-2,-2L-2,-2L-1.3333333333333335,0L-0.6666666666666667,-2Z"
              stroke="currentColor"
              stroke-linejoin="round"
              stroke-width="1"
            />
            <circle
              class="visx-circle text-gray-200"
              cx="-2"
              cy="-2"
              fill="currentColor"
              r="1"
            />
          </g>
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
      style="width: 100%; height: 100%;"
    >
      <svg
        height="0"
        width="0"
      >
        <g
          class="visx-group"
          transform="translate(1, 1)"
        >
          <path
            class="visx-area text-gray-200"
            d="M-0.6666666666666667,-2L-1.3333333333333335,0L-2,-2L-2,-2L-1.3333333333333335,0L-0.6666666666666667,-2Z"
            stroke="currentColor"
            stroke-linejoin="round"
            stroke-width="1"
          />
          <circle
            class="visx-circle text-gray-200"
            cx="-2"
            cy="-2"
            fill="currentColor"
            r="1"
          />
        </g>
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TrendLine when rendered with a positive trendline should render correcly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative"
        style="width: 100%; height: 100%;"
      >
        <svg
          height="0"
          width="0"
        >
          <g
            class="visx-group"
            transform="translate(1, 1)"
          >
            <path
              class="visx-area text-teal"
              d="M-0.6666666666666667,-2L-1.3333333333333335,-1L-2,0L-2,0L-1.3333333333333335,-1L-0.6666666666666667,-2Z"
              stroke="currentColor"
              stroke-linejoin="round"
              stroke-width="1"
            />
            <circle
              class="visx-circle text-teal"
              cx="-2"
              cy="0"
              fill="currentColor"
              r="1"
            />
          </g>
        </svg>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative"
      style="width: 100%; height: 100%;"
    >
      <svg
        height="0"
        width="0"
      >
        <g
          class="visx-group"
          transform="translate(1, 1)"
        >
          <path
            class="visx-area text-teal"
            d="M-0.6666666666666667,-2L-1.3333333333333335,-1L-2,0L-2,0L-1.3333333333333335,-1L-0.6666666666666667,-2Z"
            stroke="currentColor"
            stroke-linejoin="round"
            stroke-width="1"
          />
          <circle
            class="visx-circle text-teal"
            cx="-2"
            cy="0"
            fill="currentColor"
            r="1"
          />
        </g>
      </svg>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
